include/nbl/builtin/hlsl/central_limit_blur/box_blur.hlsl

```diff
@@ -0,0 +1,297 @@
+static const uint32_t scratchSz = arithmeticSz + broadcastSz;
+
+groupshared uint32_t scratch[ scratchSz ];
+
+template<typename T, uint16_t offset>
+struct ScratchProxy
+{
+	T get( const uint32_t ix )
+	{
+		return nbl::hlsl::bit_cast< T, uint32_t >( scratch[ ix + offset ] );
+	}
+	void set( const uint32_t ix, NBL_CONST_REF_ARG( T ) value )
+	{
+		scratch[ ix + offset ] = nbl::hlsl::bit_cast< uint32_t, T >( value );
+	}
+
+	void workgroupExecutionAndMemoryBarrier()
+	{
+		nbl::hlsl::glsl::barrier();
+	}
+};
+
+ScratchProxy<float32_t, 0> prefixSumsAccessor;
+ScratchProxy<float32_t, arithmeticSz> broadcastAccessor;
+
+struct prefix_sum_t
+{
+	using bin_op_t = nbl::hlsl::plus<float32_t>;
+	using type_t = bin_op_t::type_t;
+
+	type_t operator()( NBL_CONST_REF_ARG( type_t ) value )
+	{
+		type_t retval = nbl::hlsl::workgroup::inclusive_scan<bin_op_t, arithmeticSz>::template __call<ScratchProxy<type_t, 0> >( value, prefixSumsAccessor );
+		// we barrier before because we alias the accessors for Binop
+		prefixSumsAccessor.workgroupExecutionAndMemoryBarrier();
+		return retval;
+	}
+};
+
+prefix_sum_t workgroupPrefixSum;
+
+float32_t workgroupBroadcast( NBL_CONST_REF_ARG( float32_t ) value, uint32_t index )
+{
+	using accessor_t = ScratchProxy<float32_t, arithmeticSz>;
+	return nbl::hlsl::workgroup::Broadcast<float32_t, accessor_t>( value, broadcastAccessor, index );
+}
+
+// Todo: This spillage calculation is hacky! The lower bound of `_NBL_GLSL_EXT_BLUR_SPILLAGE_LOWER_BOUND_` is just an adhoc
+// thing which happens to work for all images until now but it could very well break for the next image.
+// This:
+// `#define LOCAL_SPILLAGE ITEMS_PER_THREAD`
+// seems to be a safe lower bound, however, it must be inefficient.
+static const uint32_t SPILLAGE_LOWER_BOUND = ( WORKGROUP_SIZE / 4 );
+
+static const uint32_t IMPL_LOCAL_SPILLAGE = ( ( arithmeticSz - 1 ) / WORKGROUP_SIZE + 1 );
+
+static const uint32_t LOCAL_SPILLAGE =
+( IMPL_LOCAL_SPILLAGE > WORKGROUP_SIZE ) ? WORKGROUP_SIZE :
+	( ( IMPL_LOCAL_SPILLAGE < WORKGROUP_SIZE ) ? SPILLAGE_LOWER_BOUND : IMPL_LOCAL_SPILLAGE );
+
+
+uint32_t DataIndex( uint32_t i ) // ?
+{
+	return ( i * gl_WorkGroupSize().x ) + nbl::hlsl::glsl::gl_LocalInvocationIndex();
+}
+
+//TODO: glsl's mod(x, y) = x - y * floor(x/y). Would prolly need to implement a glsl mod
+void BoxBlur(
+	const uint32_t ch,
+	const float32_t inRadius,
+	const uint32_t wrapMode,
+	const float32_t4 borderColor,
+	BufferAccessor textureAccessor
+) {
+	float32_t blurred = textureAccessor.get( DataIndex( i ), ch );
+
+	for( uint32_t pass = 0u; pass < PASSES_PER_AXIS; ++pass )
+	{
+		float32_t previousBlockSum = 0.f;
+
+		float32_t spill[ LOCAL_SPILLAGE ];
+		for( uint32_t i = 0u; i < LOCAL_SPILLAGE; ++i )
+		{
+			float32_t scanResult = workgroupPrefixSum( blurred ) + previousBlockSum;
+			spill[ i ] = scanResult;
+			previousBlockSum = workgroupBroadcast( spill[ i ], gl_WorkGroupSize().x - 1u );
```

Comment:
there's no need to broadcast to everyone, its enough to write out the value to memory for invocation `workgroup::SubgroupContiguousIndex()==WorkgroupSize-1` then only read it by `workgroup::SubgroupContiguousIndex()==0` (see https://github.com/Devsh-Graphics-Programming/Nabla/pull/658/files#r151329896)

You'd still need a barrier between the get and the set